File: bazaar/A_Distant_Memory.lua
---@param e NPCEventSay
function event_say(e)
    if e.message:findi("hail") then
        e.self:Say("Do you remember that item being stronger than it is?... That's because it was. Bring me four of any basic item and 500pp and I will give you a version worthy of your Rose Colored memories. Bring me four copies of any Rose Colored item and 2,000pp and I will give you exactly what your memory desires- absolute Apocrypha.")
    end
end

---@param e NPCEventTrade
function event_trade(e)
    local item_lib = require("items")
    
    local platinum = 500
    local item_id = 0
    local item_count = 0
    local reward_item_id = 0

    for i = 1, 4 do
        local inst = e.trade["item"..i]
        if inst ~= nil then
            if item_count == 0 then
                item_id = inst:GetItem():ID()
                item_count = 1
            else
                if item_id ~= inst:GetItem():ID() then
                    e.self:Say("You traded different items, I need 4 of the same item.")
                    item_lib.return_items(e.self, e.other, e.trade)
                    return
                end
                item_count = item_count+1
            end
        end
    end

    if item_count ~= 4 then
        e.self:Say("I need 4 of the same item to trade.")
        item_lib.return_items(e.self, e.other, e.trade)
        return
    end

    if item_id == 0 then
        e.self:Say("I need 4 of any item.")
        item_lib.return_items(e.self, e.other, e.trade)
        return
    end

    local item_name = eq.get_item_name(item_id)

    if item_name:find("Rose Colored") == 1 then
        platinum = 2000
        reward_item_id = item_id+100000
        if item_id < 80000 then reward_item_id = item_id+10000 end
        local apoc_name = eq.get_item_name(reward_item_id)
        if apoc_name:find("Apocryphal") ~= 1 then
            e.self:Say("I am unable to upgrade this Rose Colored item to Apocryphal.")
            item_lib.return_items(e.self, e.other, e.trade)
            return
        end
        is_rose_colored = true
    end

    if item_name:find("Apocryphal") == 1 then
        e.self:Say("I am unable to upgrade an Apocryphal item")
        item_lib.return_items(e.self, e.other, e.trade)
        return
    end

    if reward_item_id == 0 then
        reward_item_id = item_id+700000
        if item_id < 10000 then reward_item_id = item_id+70000 end
        local rose_name = eq.get_item_name(reward_item_id)
        if rose_name:find("Rose Colored") ~= 1 then
            e.self:Say("I am unable to upgrade this normal item to Rose Colored.")
            item_lib.return_items(e.self, e.other, e.trade)
            return
        end
    end

	if not item_lib.check_turn_in(e.trade, {item1 = item_id, item2 = item_id, item3 = item_id, item4 = item_id, platinum = platinum}) then
        e.self:Say(string.format("I require %d platinum to upgrade this item.", platinum))
        item_lib.return_items(e.self, e.other, e.trade)
        return
	end

    e.self:Say("Here, this is how I remember it.")
    e.other:SummonItem(reward_item_id)
    e.other:Ding()
	item_lib.return_items(e.self, e.other, e.trade)
end

File: bazaar/A_Distant_Memory.pl
sub EVENT_SAY {
    if ($text=~/hail/i) {
        quest::say("Do you remember that item being stronger than it is?... That's because it was. Bring me four of any basic item 
                    and 500pp and I will give you a version worthy of your Rose Colored memories. Bring me four copies of any 
                    Rose Colored item and 2,000pp and I will give you exactly what your memory desires- absolute Apocrypha.");
    }
}

sub EVENT_ITEM {
    my $found_item_id   = 0; 
    foreach my $item (keys %itemcount) {
        if ($item > 0) {
            $found_item_id = $item;            
            last;
        } elsif (plugin::IsItemTier2($item)) {
            quest::say("This object is already Apocryphal. I can do nothing further.");
            $found_item_id = -1;
        }
    }

    my @upgrades = plugin::GetUpgrades(plugin::GetBaseID($found_item_id));

    if ($found_item_id > 0 && (plugin::IsItemTier0($found_item_id) && !defined($upgrades[1]) || (plugin::IsItemTier1($found_item_id) && !defined($upgrades[2])))) {
        quest::say("This object has no further potential to unlock.");
        $found_item_id = -1;
    }

    if ($found_item_id >= 0) {
        if ($itemcount{$found_item_id} < 4 || $found_item_id == 0) {
            quest::say("You need to give me exactly four identical objects that you want me to upgrade.");
        } else {
            my $platinum_cost = 0;
            my $upgrade_item  = 0;
            
            if (defined($upgrades[1]) && $found_item_id < 1000000) {
                $platinum_cost = 500;
                $upgrade_item  = $upgrades[1];
            } elsif (defined($upgrades[2]) && $found_item_id < 2000000) {
                $platinum_cost = 2000;
                $upgrade_item  = $upgrades[2];
            } else {
                quest::say("I cannot upgrade this object.");
            }

            if ($upgrade_item > 0 && plugin::CheckCashPayment($platinum_cost * 1000, $copper, $silver, $gold, $platinum)) {
                quest::say("Here. Infused with greater power.");
                $client->SummonFixedItem($upgrade_item);
                return;
            } else {
                quest::say("You did not give me enough money to pay for this infusion.");
            }
        }
    }

    plugin::CheckCashPayment(0, $copper, $silver, $gold, $platinum);
    plugin::return_items(\%itemcount, 1);
}